modules/gitrepo/remote.go

```diff
@@ -0,0 +1,97 @@
+	"code.gitea.io/gitea/modules/globallock"
+	"code.gitea.io/gitea/modules/util"
+)
+
+type RemoteOption string
+
+const (
+	RemoteOptionMirrorPush  RemoteOption = "--mirror=push"
+	RemoteOptionMirrorFetch RemoteOption = "--mirror=fetch"
+)
+
+func GitRemoteAdd(ctx context.Context, repo Repository, remoteName, remoteURL string, options ...RemoteOption) error {
+	return globallock.LockAndDo(ctx, getRepoConfigLockKey(repo.RelativePath()), func(ctx context.Context) error {
+		cmd := git.NewCommand("remote", "add")
+		if len(options) > 0 {
+			switch options[0] {
+			case RemoteOptionMirrorPush:
+				cmd.AddArguments("--mirror=push")
+			case RemoteOptionMirrorFetch:
+				cmd.AddArguments("--mirror=fetch")
+			default:
+				return errors.New("unknown remote option: " + string(options[0]))
+			}
+		}
+		_, _, err := cmd.
+			AddDynamicArguments(remoteName, remoteURL).
+			RunStdString(ctx, &git.RunOpts{Dir: repoPath(repo)})
+		return err
+	})
+}
+
+func GitRemoteRemove(ctx context.Context, repo Repository, remoteName string) error {
+	return globallock.LockAndDo(ctx, getRepoConfigLockKey(repo.RelativePath()), func(ctx context.Context) error {
+		cmd := git.NewCommand("remote", "rm").AddDynamicArguments(remoteName)
+		_, _, err := cmd.RunStdString(ctx, &git.RunOpts{Dir: repoPath(repo)})
+		return err
+	})
+}
+
+// GitRemoteGetURL returns the url of a specific remote of the repository.
+func GitRemoteGetURL(ctx context.Context, repo Repository, remoteName string) (*giturl.GitURL, error) {
+	addr, err := git.GetRemoteAddress(ctx, repoPath(repo), remoteName)
+	if err != nil {
+		return nil, err
+	}
+	if addr == "" {
+		return nil, util.NewNotExistErrorf("remote '%s' does not exist", remoteName)
+	}
+	return giturl.ParseGitURL(addr)
+}
+
+func SetRemoteURL(ctx context.Context, repo Repository, remoteName, remoteURL string) error {
+	return globallock.LockAndDo(ctx, getRepoConfigLockKey(repo.RelativePath()), func(ctx context.Context) error {
+		cmd := git.NewCommand("remote", "set-url").AddDynamicArguments(remoteName, remoteURL)
+		_, _, err := cmd.RunStdString(ctx, &git.RunOpts{Dir: repoPath(repo)})
+		return err
+	})
+}
+
+// GitRemotePrune prunes the remote branches that no longer exist in the remote repository.
+// No lock is needed because the remote remoteName will be checked before invoking this function.
+// Then it will not update the remote automatically if the remote does not exist.
+func GitRemotePrune(ctx context.Context, repo Repository, remoteName string, timeout time.Duration, stdout, stderr io.Writer) error {
+	return git.NewCommand("remote", "prune").AddDynamicArguments(remoteName).
+		Run(ctx, &git.RunOpts{
+			Timeout: timeout,
+			Dir:     repoPath(repo),
+			Stdout:  stdout,
+			Stderr:  stderr,
+		})
+}
+
+// GitRemoteUpdatePrune updates the remote branches and prunes the ones that no longer exist in the remote repository.
+// No lock is needed because the remote remoteName will be checked before invoking this function.
```

Comment:
Where is the "remoteName will be checked"?

Reply:
b40e8fd488d19518293d4c5da66a993c04be55df